File: database/database-annotation-processor/src/main/java/ru/tinkoff/kora/database/annotation/processor/r2dbc/R2dbcNativeTypes.java
package ru.tinkoff.kora.database.annotation.processor.r2dbc;

import com.squareup.javapoet.ArrayTypeName;
import com.squareup.javapoet.ClassName;
import com.squareup.javapoet.TypeName;

import javax.annotation.Nullable;
import java.math.BigDecimal;
import java.math.BigInteger;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.Set;

public class R2dbcNativeTypes {

    private static final Set<TypeName> nativeTypes = Set.of(
        TypeName.INT,
        TypeName.INT.box(),
        TypeName.LONG,
        TypeName.LONG.box(),
        TypeName.DOUBLE,
        TypeName.DOUBLE.box(),
        TypeName.BOOLEAN,
        TypeName.BOOLEAN.box(),
        ClassName.get(String.class),
        ClassName.get(BigDecimal.class),
        ClassName.get(BigInteger.class),
        ClassName.get(LocalDateTime.class),
        ClassName.get(LocalDate.class),
        ArrayTypeName.of(TypeName.BYTE)
    );

    @Nullable
    public static TypeName findAndBox(TypeName typeName) {
        if (nativeTypes.contains(typeName)) {
            return typeName.isPrimitive()
                ? typeName.box()
                : typeName;
        }
        return null;
    }
}

File: database/database-symbol-processor/src/main/kotlin/ru/tinkoff/kora/database/symbol/processor/r2dbc/R2dbcNativeTypes.kt
package ru.tinkoff.kora.database.symbol.processor.r2dbc

import com.squareup.kotlinpoet.TypeName
import com.squareup.kotlinpoet.asTypeName
import java.math.BigDecimal
import java.math.BigInteger
import java.time.LocalDate
import java.time.LocalDateTime

object R2dbcNativeTypes {
    val types = listOf(
        R2dbcNativeType.of(Boolean::class.asTypeName()),
        R2dbcNativeType.of(Boolean::class.asTypeName().copy(true)),
        R2dbcNativeType.of(Int::class.asTypeName()),
        R2dbcNativeType.of(Int::class.asTypeName().copy(true)),
        R2dbcNativeType.of(Long::class.asTypeName()),
        R2dbcNativeType.of(Long::class.asTypeName().copy(true)),
        R2dbcNativeType.of(Double::class.asTypeName()),
        R2dbcNativeType.of(Double::class.asTypeName().copy(true)),
        R2dbcNativeType.of(String::class.asTypeName()),
        R2dbcNativeType.of(String::class.asTypeName().copy(true)),
        R2dbcNativeType.of(BigDecimal::class.asTypeName()),
        R2dbcNativeType.of(BigDecimal::class.asTypeName().copy(true)),
        R2dbcNativeType.of(BigInteger::class.asTypeName()),
        R2dbcNativeType.of(BigInteger::class.asTypeName().copy(true)),
        R2dbcNativeType.of(LocalDateTime::class.asTypeName()),
        R2dbcNativeType.of(LocalDateTime::class.asTypeName().copy(true)),
        R2dbcNativeType.of(LocalDate::class.asTypeName()),
        R2dbcNativeType.of(LocalDate::class.asTypeName().copy(true)),
        R2dbcNativeType.of(ByteArray::class.asTypeName()),
        R2dbcNativeType.of(ByteArray::class.asTypeName().copy(true)),
    )


    fun findNativeType(typeName: TypeName) = types.firstOrNull { it.type() == typeName }
}
